Product and Process Support - Check PR

battle-panda opened the pull request:
# Purpose
The purpose of this change is to implement the `hyaline check pr`.

- https://github.com/appgardenstudios/hyaline/issues/202

# Changes
- Implemented `hyaline check PR` using diffs from the PR
- Added launch config
- Added e2e tests

# Testing
- Run e2e and unit tests
- Use launch configs to manually test

compiledpanda commented on cli/internal/action/checkPR.go:
```diff
@@ -0,0 +1,487 @@
+	}
+
+	// Output the comment to a file
+	jsonData, err := json.MarshalIndent(comment, "", "  ")
```
Lets use the same struct for the output of check diff and check pr. That way any downstream tooling can operate on either one interchangeably.

battle-panda replied:
@compiledpanda An implication of how I have implemented things with my most recent changes is that only recommendations that we found on our current run will be included in the output. In other words, we will not include recommendations that were in an existing PR comment that we merged with when updating the comment.

I think this is okay for the following reasons, but I'd like to get your thoughts:
- Although some people might expect to get the recommendations from a previous run, I think most people would expect only recommendations from the current run
- There is and will potentially be context about existing recommendations that we currently don't have a place in this output format to represent. For example, recommendations that have been removed, or checked off.
- In theory, the user could have modified the PR comment and included things on their own, which I don't know if we would want to include
- I imagine we might want to include a flag at some point to disable the PR-comment functionality and to only output in the file. In which case, we wouldn't have historical data anyway (although that could just be considered a consequence of using the disable flag)

Overall, this does highlight some constraints or considerations that we are introducing by coupling the output formats. I _think_ it's still the right decision, but it's something to be aware of. Maybe we would introduce a secondary output flag in the future to "output the comment" (similar to what updatePR used to do).

Thoughts?

battle-panda replied:
On a related note, I'm planning on making `--output` optional, since I think most of the time, people will probably just want the PR comment updated

compiledpanda commented on .vscode/launch.json:
```diff
@@ -345,6 +345,42 @@
+        "--pull-request", "appgardenstudios/hyaline-example/9",
+        "--issue", "appgardenstudios/hyaline-example/2",
+        "--issue", "appgardenstudios/hyaline-example/3",
+        "--output", "../.vscode/hyaline/check-pr/comment.json"
```
Going along with the comment about making the output recommendations in both check commands, lets call the output recommendations.json like the output of check diff.

compiledpanda commented on cli/cmd/hyaline/check.go:
```diff
@@ -230,6 +230,55 @@ func Check(logLevel *slog.LevelVar) *cli.Command {
+					&cli.StringFlag{
+						Name:     "output",
+						Required: true,
+						Usage:    "Path to write the results to",
```
We should probably say recommendations instead of results here.

compiledpanda commented on cli/internal/action/checkPR.go:
```diff
@@ -0,0 +1,487 @@
+}
+
+type CheckPRComment struct {
+	Sha             string                         `json:"sha"`
```
Along with the comment about matching outputs between the check commands, we could also update the check output to include the sha in both places.

And riffing off of that idea, we could extend this to return both the base and head sha used in the code diff for recommendations.

compiledpanda commented on cli/internal/action/checkPR.go:
```diff
@@ -0,0 +1,487 @@
+	return fmt.Sprintf("<![CDATA[ %s ]]>", string(data)), nil
+}
+
+const CHECKPR_HYALINE_HEADER = "# H\u200By\u200Ba\u200Bl\u200Bi\u200Bn\u200Be"
```
nit: It might be good to move these consts to the top of the file.

compiledpanda commented on cli/internal/check/diff.go:
```diff
@@ -146,18 +154,29 @@ func formatCheckPrompt(file code.FilteredFile, documents []*docs.FilteredDoc, pr
-		// Add prompt
-		prompt.WriteString(fmt.Sprintf("Given that the file %s was created, ", file.Filename))
-		prompt.WriteString("and that the contents of the created file are in <file>, ")
+		if file.Diff != "" {
```
We may want to flip this logic here and below so that we use the file contents if we have them, and then use the diff if we don't. My reasoning is that we will use fewer tokens by posting the contents of the file than the diff, as well as the diff in this case feels unnecessary as long as we have the contents.

battle-panda commented on cli/internal/code/getFilteredPRFiles.go:
```diff
@@ -0,0 +1,131 @@
+		case "added":
+			changedFiles[filename] = struct{}{}
+			if config.PathIsIncluded(filename, cfg.Include, cfg.Exclude) {
+				diff := fmt.Sprintf("--- /dev/null\n+++ b/%s\n", filename)
```
This is the format that git uses. Example:

```bash
$ git diff b43fcf50cffc6468a6725a360abc739ad16d37b7 6aed98c5d3bcfd5139556910aa809c8ccd3efb79
diff --git a/subdir/test.js b/subdir/test.js
new file mode 100644
index 0000000..b7b1ddb
--- /dev/null
+++ b/subdir/test.js
@@ -0,0 +1 @@
+console.log("This is a test file in a subdirectory.");
```

battle-panda commented on cli/internal/code/getFilteredPRFiles.go:
```diff
@@ -0,0 +1,131 @@
+				}
+			}
+		default:
+			slog.Debug("code.GetFilteredPRFiles unknown file status", "status", *file.Status, "filename", filename)
```
I actually disagree. This is in part to cover the status of "unmodified" we are ignoring. Although, I could make that explicit. However, if GitHub introduces a new status type, I don't think everyone who is using hyaline should have it blow up at them.

battle-panda replied:
Explicitly handle the ones we are aware of (even if the handling is to skip it). In the default case, do a slog.Warn indicating that it is unexpected and a request to report.

battle-panda commented:
@compiledpanda I believe I have addressed all PR feedback and it is ready for another look, although it would be good to get your thoughts on this: https://github.com/appgardenstudios/hyaline/pull/229#discussion_r2255334644

compiledpanda commented on cli/internal/action/checkDiff.go:
```diff
@@ -29,41 +32,36 @@ type CheckDiffArgs struct {
-type CheckDiffRecommendation struct {
-	Source         string   `json:"documentationSource"`
-	Document       string   `json:"document"`
-	Section        []string `json:"section,omitempty"`
```
Let's talk about this change IRL. I _think_ we want to omit empty here still?

compiledpanda replied:
Per an IRL conversation, we'll remove the omitempty so that section is always present.

compiledpanda commented on cli/internal/action/checkPR.go:
```diff
@@ -0,0 +1,459 @@
+		return false
+	}
+
+	return reflect.DeepEqual(newRec.Section, existingRec.Section)
```
Per a conversation IRL let's use a join and compare here
